### PR TITLE
Buildings catching water

### PR DESCRIPTION
**Pull request by capttrousers**
issue #29 
buildings catching water problem, using recursion

### CODE REVIEW

File: src/test/java/BuildingsCatchingWaterTest.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        Assert.assertEquals(expectedWaterCaught, BuildingsCatchingWater.compute(buildings));
+    }
+
+    @Test
+    public void testCase5() {
+        int[] buildings = new int[]{3, 1, 2, 1, 2, 4, 1, 2, 3, 5};
+        int   expectedWaterCaught = 12;
+        Assert.assertEquals(expectedWaterCaught, BuildingsCatchingWater.compute(buildings));
+    }
+}
```
Comment (shalka) [inline src/test/java/BuildingsCatchingWaterTest.java:135]: add tests for the negative cases (null, empty array)

File: src/main/java/BuildingsCatchingWater.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        while(left < right && buildings[right - 1] >= buildings[right]) {
+            right--;
+        }
+        if(left == right) {
+            return 0;
+        }
+        return BCW(buildings, left, right);
+    }
+
+    private static int BCW(int[] buildings, int left, int right) {
```
Comment (shalka) [inline src/main/java/BuildingsCatchingWater.java]: this is a good recursive implementation, but it's going to get expensive if larger arrays are passed in. can we do this incrementally, without recursion?

File: src/main/java/BuildingsCatchingWater.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        // find building C that is tallest building between left and right
+        int maxHeight = 0;
+        int maxHeightIndex = left + 1;
+        for(int i = maxHeightIndex; i < right; i++){
+            if(buildings[i] > maxHeight) {
+                maxHeightIndex = i;
+                maxHeight = buildings[i];
+            }
+        }
+        return ((right - left - 1) * (minHeight - maxHeight)) + BCW(buildings, left, maxHeightIndex) + BCW(buildings, maxHeightIndex, right);
```
Comment (shalka) [inline src/main/java/BuildingsCatchingWater.java]: the inlined logic in the return statement should be broken up so that its more understandable.

**Comment by shalka** [comment]
looks great! :shipit:
